### PR TITLE
Adding lepton in fatjet vars

### PR DESCRIPTION
**Pull request by cmantill**
#### PR description:

Adding LSF variable to be used by B2G and other (e.g. EXO) analysis. (e.g. @pmaksim1 as head of B2G has expressed interested, as presented here: https://indico.cern.ch/event/871916/contributions/3677160/attachments/1969064/3274986/LeptonIsolation_B2G_14Jan.pdf)

LSF stands for lepton subjet fraction https://arxiv.org/pdf/1410.0362.pdf and is currently using PF electrons and muons. It can be used to identify boosted decays of jets with a close lepton (e.g. WW => lnuqq).
The algorithm reclusters the PF constituents of each fat jet (AK8) into n subjets using the exclusive kT algorithm, with n optimally chosen to match the number of
subjets expected in signal events (in this case n=3). Once this is done, all leptons in the event have been associated with a subjet. The lepton subjet fraction of each lepton is defined as the ratio of the PF lepton pT to its associated subjet pT. Leptons with higher LSF values are considered to be more isolated.

It adds the following vars to AK8 jets:
lsf3, electron and muon indices matched to the AK8 jet (-1 by default)

#### PR validation:

This implementation has been tested locally with TT sample.

#### if this PR is a backport please specify the original PR:

This PR was already presented last year and the original code went through some initial review in this repo:
https://github.com/cms-nanoAOD/cmssw/pull/299
The PR was then closed for lack of activity.

### CODE REVIEW

**Comment by clelange** [comment]
This would be something great to have for nanoAOD v7 already. Can this please be reviewed?

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 1]
```diff
@@ -155,12 +154,22 @@
     looseId = cms.InputTag("looseJetId"),
 )
 
+lepInJetVars = cms.EDProducer("LepInJetProducer",
+    src = cms.InputTag("updatedJetsAK8"),
+    srcEle = cms.InputTag("slimmedElectrons"),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/jets_cff.py]: I start the test already, but this is incorrect.
Not all elements of the slimmedElectrons and slimmedMuons are saved, so the indices will be invalid once read in NanoAOD. Please adapt to use the finalElectrons and finalMuons collections.

**Review by peruzzim — CHANGES_REQUESTED** [review]

**Comment by peruzzim** [comment]
@cmantill please follow up on the comment above and apply code-format and DQM updates

File: PhysicsTools/NanoAOD/python/jets_cff.py (hunk 1 above)
Comment (cmantill) [reply PhysicsTools/NanoAOD/python/jets_cff.py]: Ok I just made the changes to finalElectrons and finalMuons

**Review by cmantill — COMMENTED** [review]
Ok, just modified DQM file, code format and changes to lepton collections.

**Comment by cmantill** [comment]
@peruzzim Sorry, I'm a bit confused, from the log file it seems that finalMuons comes after than jets: https://github.com/cms-nanoAOD/cmssw/blob/master-cmsswmaster/PhysicsTools/NanoAOD/python/nano_cff.py#L125 

Should I try to modify the nanoSequence or keep the slimmedMuons and Electrons as input?

**Comment by peruzzim** [comment]
Exactly, the indices must stay aligned, I'm afraid we need to place the module in a sequence to run after the jet and muon ones, and read them as an `ExtVar` in the `fatJetTable` so you don't have to add them as `userFloat` (this means reading directly from a value map, see e.g. what is done for mvaTTH with muons).
Could you give this a try, please?

**Comment by cmantill** [comment]
@peruzzim Something like this? https://github.com/cms-nanoAOD/cmssw/pull/484/commits/3a73fcc1913ef2796e5035bfc4693cb6cceabbdf It seems to run locally for me but one thing I'm not sure is whether the jets source should be: `src = cms.InputTag("finalJetsAK8"),` or something like `src = cms.InputTag("linkedObjects","fatjets")` ?

**Comment by peruzzim** [comment]
Thank you, I have added a couple of commits to move the module in a separate sequence and avoid mixing with the part that fills tables.

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 2]
```diff
@@ -178,7 +176,11 @@
     cut = cms.string("pt > 170")
 )
 
-
+lepInJetVars = cms.EDProducer("LepInJetProducer",
+    src = cms.InputTag("finalJetsAK8"),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/jets_cff.py]: this needs to be `updatedJetsAK8WithUserData`

File: PhysicsTools/NanoAOD/plugins/LeptonInJetProducer.cc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  edm::Handle<edm::View<pat::Electron>> srcEle;
+  iEvent.getByToken(srcEle_, srcEle);
+  edm::Handle<edm::View<pat::Muon>> srcMu;
+  iEvent.getByToken(srcMu_, srcMu);
+
+  unsigned int nJet = srcJet->size();
+  unsigned int nEle = srcEle->size();
+  unsigned int nMu = srcMu->size();
+
+  std::vector<float> *vlsf3 = new std::vector<float>;
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/plugins/LeptonInJetProducer.cc]: these look like memory leaks, please use a simple objects like `std::vector<float> vlsf3` or unique_ptr

**Review by peruzzim — CHANGES_REQUESTED** [review]

**Comment by cmantill** [comment]
@peruzzim Ok thank you, I have added the commits above to address your comments

**Review by peruzzim — APPROVED** [review]

**Comment by pmaksim1** [comment]
Thanks, Marco!

**Pull request merged**